### DISCUSSION TITLE
dump(): incorrect precision for float array

### DISCUSSION THREAD

**Discussion by yhmo** [nlohmann/json#2254] [General]
#### What is the issue you have?

I am using nlohmann::json version 0.3.7 in my project, it is a great tool.

But I encounter a small problem:  dump a json object which contains float array, seems the precision is incorrect.

#### Can you provide a small but working code example?

```
#include <vector>
#include <iostream>

#include "json.hpp"

int main() {
    std::vector<float> float_data = {0.1234567, 4.123456};
    nlohmann::json json;
    json["AAA"] = float_data;
    std::string ss = json.dump();
    std::cout << ss << std::endl;
}
```

But the output string is:
{"AAA":[0.12345670163631439,4.123456001281738]}

#### What is the expected behavior?

I suppose the output string should be:
{"AAA":[0.1234567,4.123456]}

#### Which version of the library did you use?

- [x] latest release version 3.7.0

**Comment by nlohmann** [comment]
The library uses `double` as default type for floating-point numbers. Therefore, your numbers are converted to `double` first and then serialized using as many digits as necessary to roundtrip `double`s.

If you want to use `float` instead, you can override the default type templates this:

```cpp
#include <iostream>
#include "json.hpp"

int main()
{
    std::vector<float> float_data = {0.1234567, 4.123456};
    nlohmann::basic_json<std::map, std::vector, std::string, bool, std::int64_t, std::uint64_t, float> json;
    json["AAA"] = float_data;
    std::string ss = json.dump();
    std::cout << ss << std::endl;
}
```

Output:

```json
{"AAA":[0.1234567,4.123456]}
```

**Reply by yhmo** [reply]
It works. Thank you!
